File: src/main/java/com/netsensia/rivalchess/model/SquareOccupant.java
package com.netsensia.rivalchess.model;

import com.netsensia.rivalchess.model.exception.EnumConversionException;

public enum SquareOccupant {
    NONE(-1),
    WP(0),
    WN(1),
    WB(2),
    WQ(3),
    WK(4),
    WR(5),
    BP(6),
    BN(7),
    BB(8),
    BQ(9),
    BK(10),
    BR(11)
    ;

    private int index;

    private SquareOccupant(int index) {
        this.index = index;
    }

    public int getIndex() {
        return index;
    }

    public SquareOccupant ofColour(Colour colour) {
        if (this == SquareOccupant.NONE) {
            return SquareOccupant.NONE;
        }

        return colour == Colour.WHITE
                ? SquareOccupant.fromIndex(index % 6)
                : SquareOccupant.fromIndex(index % 6 + 6);
    }

    private static SquareOccupant fromIndex(int index){
        for(SquareOccupant cp : SquareOccupant.values()){
            if(cp.index == index) {
                return cp;
            }
        }
        return SquareOccupant.NONE;
    }

    public Colour getColour() {
        if (index == SquareOccupant.NONE.getIndex()) {
            throw new EnumConversionException("Can't get piece colour of an empty square");
        }

        return index <= SquareOccupant.WR.getIndex() ? Colour.WHITE : Colour.BLACK;
    }

    public Piece getPiece() {
        switch (this) {
            case NONE:
                return Piece.NONE;
            case WP:
            case BP:
                return Piece.PAWN;
            case WK:
            case BK:
                return Piece.KING;
            case WR:
            case BR:
                return Piece.ROOK;
            case WB:
            case BB:
                return Piece.BISHOP;
            case WN:
            case BN:
                return Piece.KNIGHT;
            case WQ:
            case BQ:
                return Piece.QUEEN;
            default:
                throw new EnumConversionException("Invalid SquareOccupant");
        }
    }

    public char toChar() {
        switch (this) {
            case NONE:
                return '-';
            case WP:
                return 'P';
            case BP:
                return 'p';
            case WK:
                return 'K';
            case BK:
                return 'k';
            case WR:
                return 'R';
            case BR:
                return 'r';
            case WB:
                return 'B';
            case BB:
                return 'b';
            case WN:
                return 'N';
            case BN:
                return 'n';
            case WQ:
                return 'Q';
            case BQ:
                return 'q';
            default:
                throw new EnumConversionException("Invalid SquareOccupant");
        }
    }

    public static SquareOccupant fromChar(char piece){
        switch (piece) {
            case 'p':
                return SquareOccupant.BP;
            case 'q':
                return SquareOccupant.BQ;
            case 'r':
                return SquareOccupant.BR;
            case 'n':
                return SquareOccupant.BN;
            case 'b':
                return SquareOccupant.BB;
            case 'k':
                return SquareOccupant.BK;
            case 'P':
                return SquareOccupant.WP;
            case 'Q':
                return SquareOccupant.WQ;
            case 'R':
                return SquareOccupant.WR;
            case 'N':
                return SquareOccupant.WN;
            case 'B':
                return SquareOccupant.WB;
            case 'K':
                return SquareOccupant.WK;
            default:
                return SquareOccupant.NONE;
        }
    }

}

File: src/main/java/com/netsensia/rivalchess/model/SquareOccupant.kt
package com.netsensia.rivalchess.model

import com.netsensia.rivalchess.model.exception.EnumConversionException

enum class SquareOccupant(val index: Int) {
    NONE(-1), WP(0), WN(1), WB(2), WQ(3), WK(4), WR(5), BP(6), BN(7), BB(8), BQ(9), BK(10), BR(11);

    fun ofColour(colour: Colour?): SquareOccupant {
        return when (colour) {
            Colour.WHITE -> when (this) {
                WP, BP -> WP
                WR, BR -> WR
                WQ, BQ -> WQ
                WB, BB -> WB
                WK, BK -> WK
                WN, BN -> WN
                else -> NONE
            }
            Colour.BLACK -> when (this) {
                WP, BP -> BP
                WR, BR -> BR
                WQ, BQ -> BQ
                WB, BB -> BB
                WK, BK -> BK
                WN, BN -> BN
                else -> NONE
            }
            else -> throw EnumConversionException("Invalid colour")
        }
    }

    val colour: Colour
        get() = when (this) {
            NONE -> throw EnumConversionException("CCan't get piece colour of an empty square")
            WP, WK, WN, WB, WR, WQ -> Colour.WHITE
            BP, BK, BN, BB, BR, BQ -> Colour.BLACK
        }

    val piece: Piece
        get() = when (this) {
            NONE -> Piece.NONE
            WP, BP -> Piece.PAWN
            WK, BK -> Piece.KING
            WR, BR -> Piece.ROOK
            WB, BB -> Piece.BISHOP
            WN, BN -> Piece.KNIGHT
            WQ, BQ -> Piece.QUEEN
        }

    fun toChar(): Char {
        return when (this) {
            NONE -> '-'
            WP -> 'P'
            BP -> 'p'
            WK -> 'K'
            BK -> 'k'
            WR -> 'R'
            BR -> 'r'
            WB -> 'B'
            BB -> 'b'
            WN -> 'N'
            BN -> 'n'
            WQ -> 'Q'
            BQ -> 'q'
            else -> throw EnumConversionException("Invalid SquareOccupant")
        }
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun fromChar(piece: Char): SquareOccupant {
            return when (piece) {
                'p' -> BP
                'q' -> BQ
                'r' -> BR
                'n' -> BN
                'b' -> BB
                'k' -> BK
                'P' -> WP
                'Q' -> WQ
                'R' -> WR
                'N' -> WN
                'B' -> WB
                'K' -> WK
                else -> NONE
            }
        }
    }

}